### PR TITLE
Fix level db panic

### PR DESCRIPTION
**Pull request by michaelavila**
This PR exists for me to trouble shoot some CI issues. Disregard.

### CODE REVIEW

**Comment by Stebalien** [comment]
https://github.com/ipfs/go-ipfs/pull/6185?

**Comment by michaelavila** [comment]
@Stebalien I'm actually using this PR to troubleshoot the leveldb panics as I can't reproduce them locally. Are you saying #6185 should fix that?

**Comment by Stebalien** [comment]
Ah, no. Sorry, different issue.

In terms of the leveldb panics, I _think_ we're closing the datastore while reading from it (or reading from it after closing it?).

**Comment by michaelavila** [comment]
@Stebalien yeah, there are those panics and there's also the panic from the pinning test:

```
panic: interface conversion: cache.Value is nil, not *table.Reader

goroutine 103 [running]:
github.com/syndtr/goleveldb/leveldb.(*tOps).find(0xc00034fc20, 0xc0001bc0f0, 0xc000456090, 0x2b, 0x2b, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/circleci/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.0/leveldb/table.go:383 +0x221
github.com/syndtr/goleveldb/leveldb.(*version).get.func1(0x0, 0xc0001bc0f0, 0xc000456090)
	/home/circleci/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.0/leveldb/version.go:176 +0x452
github.com/syndtr/goleveldb/leveldb.(*version).walkOverlapping(0xc000330a50, 0x0, 0x0, 0x0, 0xc000456090, 0x2b, 0x2b, 0xc000347910, 0xc0003478e0)
	/home/circleci/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.0/leveldb/version.go:116 +0x2a7
github.com/syndtr/goleveldb/leveldb.(*version).get(0xc000330a50, 0x0, 0x0, 0x0, 0xc000456090, 0x2b, 0x2b, 0x0, 0x0, 0x0, ...)
	/home/circleci/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.0/leveldb/version.go:160 +0x2b2
github.com/syndtr/goleveldb/leveldb.(*DB).get(0xc000286000, 0x0, 0x0, 0x0, 0x0, 0xc000456060, 0x23, 0x30, 0x1d, 0x0, ...)
	/home/circleci/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.0/leveldb/db.go:779 +0x397
github.com/syndtr/goleveldb/leveldb.(*DB).Get(0xc000286000, 0xc000456060, 0x23, 0x30, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/circleci/go/pkg/mod/github.com/syndtr/goleveldb@v1.0.0/leveldb/db.go:845 +0x12f
github.com/ipfs/go-ds-leveldb.(*accessor).Get(0xc000190a20, 0xc000456030, 0x23, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/circleci/go/pkg/mod/github.com/ipfs/go-ds-leveldb@v0.0.1/datastore.go:84 +0x82
github.com/ipfs/go-ds-measure.(*measure).Get(0xc0001de000, 0xc000456030, 0x23, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/circleci/go/pkg/mod/github.com/ipfs/go-ds-measure@v0.0.1/measure.go:149 +0xd4
github.com/ipfs/go-datastore/mount.(*Datastore).Get(0xc0001942c0, 0xc000456000, 0x23, 0x24bcc00, 0xbf21969c66d5cebf, 0x1336e5a, 0x249da20, 0x23)
	/home/circleci/go/pkg/mod/github.com/ipfs/go-datastore@v0.0.2/mount/mount.go:197 +0x70
github.com/ipfs/go-ds-measure.(*measure).Get(0xc0001de240, 0xc000456000, 0x23, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/circleci/go/pkg/mod/github.com/ipfs/go-ds-measure@v0.0.1/measure.go:149 +0xd4
github.com/ipfs/go-datastore/keytransform.(*ktds).Get(0xc000336400, 0xc0001ac040, 0x11, 0x1, 0xc0001ae090, 0x10, 0x14a5580, 0x24bdfe0)
	/home/circleci/go/pkg/mod/github.com/ipfs/go-datastore@v0.0.2/keytransform/keytransform.go:40 +0x76
github.com/ipfs/go-ipfs/provider.(*Queue).nextEntry(0xc0004d8a50, 0x0, 0x0, 0xc00009d601, 0xc00009d698)
	/home/circleci/ipfs/go-ipfs/provider/queue.go:73 +0xc2
github.com/ipfs/go-ipfs/provider.(*Queue).work.func1(0xc0004d8a50)
	/home/circleci/ipfs/go-ipfs/provider/queue.go:107 +0x6d1
created by github.com/ipfs/go-ipfs/provider.(*Queue).work
	/home/circleci/ipfs/go-ipfs/provider/queue.go:101 +0x4f
```

They could be related. The other panics (the `10:56:51.864 ERROR   provider: Failed to enqueue cid: leveldb: closed queue.go:121`) are just being logged and swallowed up, but the panic in the middle of the pinning test is different and actually panicking `ipfs`, so I'm a little more worried about it.

File: provider/queue.go [hunk 1]
```diff
@@ -36,19 +37,29 @@ func NewQueue(ctx context.Context, name string, ds datastore.Datastore) (*Queue,
...
+		close:   cancel,
 	}
 	q.work()
 	return q, nil
 }
 
+// Close stops the queue
+func (q *Queue) Close() error {
+	log.Error("Before close")
+	q.close()
```
Comment (Stebalien) [inline provider/queue.go]: We may want to wait for the queue to actually close (e.g., have some channel to signal that it's done).

File: provider/queue.go [hunk 2]
```diff
@@ -44,11 +46,24 @@ func NewQueue(ctx context.Context, name string, ds datastore.Datastore) (*Queue,
...
+		close:   make(chan struct{}),
+		done:    make(chan struct{}, 1),
 	}
 	q.work()
 	return q, nil
 }
 
+// Close stops the queue
+func (q *Queue) Close() error {
+	for {
```
Comment (Stebalien) [inline provider/queue.go]: Probably simpler to just cancel the context and then wait on `q.done`.

File: provider/queue.go [hunk 3]
```diff
@@ -102,6 +117,10 @@ func (q *Queue) work() {
 		var k datastore.Key = datastore.Key{}
 		var c cid.Cid = cid.Undef
 
+		defer func() {
+			q.done <- struct{}{}
```
Comment (Stebalien) [inline provider/queue.go]: The idiomatic way to do this is to close the channel. That way, all future reads will return immediately.

**Comment by michaelavila** [comment]
@Stebalien we had something like your suggestion earlier, but ran into some issues that I didn't this time around. Changes are in. Let me know if you had something else in mind.

**Comment by Stebalien** [comment]
Ok, one more blocker down. Thanks!

**Pull request merged**